### PR TITLE
ssh: Eliminate $COCKPIT_SSH_KNOWN_HOSTS_DATA

### PR DESCRIPTION
**Pull request by martinpitt**
* [x] Implementation
 * [x] Update tests
 * [x] Ensure we have tests for whatever old or compat behavior we still carry.
 * [x] Check the [ManageIQ](https://github.com/ManageIQ/manageiq) code base to make sure this does not break functionality.
 * [x] Check the  [ovirt-cockpit-sso](https://github.com/oVirt/ovirt-cockpit-sso) to make sure this does not break functionality.
 * [X] Ensure we do not land this change in any LTS or stable operating system (such as RHEL 7.x) - Yes, this is not aimed at RHEL/CentOS 7.

### CODE REVIEW

**Comment by Gundersanne** [comment]
As I removed the ignore_host config value  overwriting localhost, the test fail because they depend on this overwrite. So adding needs-work again.

**Comment by Gundersanne** [comment]
I'm not sure we can drop `knownhosts_authorize`.

During login we should never challenge for known host data, and this `ssh_option` also controlled that, and this pr always challenges for known host data as a last option. We could maybe simplify this and get rid of the environment variable `COCKPIT_SSH_KNOWN_HOSTS_DATA` though, not sure.

 Or we need to figure out a clever way to detect that we're on the login page, and not in an authorized session already. The latter is probably not the possible.

Either way we should document this a lot better why this is so complicated.

**Comment by martinpitt** [comment]
@Gundersanne : IIRC `allow_unknown`, aka `connect_unknown_hosts` would be needed precisely for covering the "not in session" case, i. e. for direct `http://host:9090/=remotehost/` URLs -- then it would be false, and in every other case (including interactive login) it would be true.

But this seems to be *exactly* the same situation with how you described `knownhosts_authorize` - i. e. is there really a situation when `knownhosts_authorize` is not equal to `!connect_unknown_hosts`?

**Comment by Gundersanne** [comment]
> is there really a situation when knownhosts_authorize is not equal to !connect_unknown_hosts?

1. Counter-intuitively in the login case it would be `if (connect_unknown_host)` with**out** the `!`. Since that would mean we never challenge for this data during the login, yet still get the `unknown-host` problem.

2. But we might need to challenge for it during for instance the openshift-click-on-node-address-and-open-cockpit case in the integration tests which happens via `cockpitpeer.c`, and there we'd need `!connect_unknown_hosts` (with the `!`). Or when connecting to other machines via for instance the dashboard. In these cases `connect_unknown_host` could be `false` or `true`.

Now other logic depend on this value as well, so we can't just flip it. We need to be able to ignore the `unknown-host` problem independently of challenging for known host data, which is where authorize came in.

**Comment by martinpitt** [comment]
I sent PR #10192 to drop the var from the manifest. Ideally this should really be a no-op, but it might expose some unexpected test failures already, which are then easier to investigate.

**Comment by martinpitt** [comment]
Saving IRC conversation to keep this at a more permanent place.
```
14:35:37 Gundersanne | 10192 effectively disables default authorize
14:36:06       pitti | Gundersanne: if it would do that, wouldn't a *lot* more tests fail then?
14:36:39 Gundersanne | well i think a lot of tests don't depend on the authorize to succeed (note that authorize just means being challenged for known hosts data)
14:36:51       pitti | FAOD, 10192 isn't *meant* to disable default auhorize, just removing setting it twice
14:38:16 Gundersanne | right but it does disable it by default, since the way it's enabled is comparing the value of _DATA with the constant "authorize"
14:38:20       pitti | oh, I see (I think)
14:38:25 Gundersanne | if the _DATA is empty, then it's always off
14:38:35 Gundersanne | now only very few cases actually depend on authorize
14:39:06 Gundersanne | but in checkDirectLogin in multimachines it's explicitely disabled, even in master. And i don't see how
14:39:08       pitti | cockpit_ssh_options_from_env() first looks at _DATA, sets ignore_hostkey based on that, but doesn't internally default to authorize
14:39:25       pitti | I think I misunderstood the usage of const gchar *authorize_knownhosts_data = "authorize";
14:39:32       pitti | in which case the fix should be trivial
14:39:42 Gundersanne | right first it compares _DATA to "*" then to "authorize"
14:39:58       pitti | i. e. cockpit_ssh_options_from_env should just set it to authorize_knownhosts_data if it's not set
14:39:59 Gundersanne | "*" means blanket let everything through, just test purposes. "authorize" means challenge for host data pre-connect
14:40:36       pitti | so at least now I have a clear and understandable thinko :)
14:41:04 Gundersanne | well it's a boolean, so if _DATA=authorize, ssh_options->knownhosts_auhtorize is true; otherwise false
14:41:11 Gundersanne | yea it's not easy to reason about
14:41:25           * | Gundersanne has been doing nothing else for a week now :1
14:41:32       pitti | Gundersanne: so WDYT, I fix 10192 to *actually* make it the internal default
14:41:57       pitti | and then in 10143 we can see what's left there
14:42:48 Gundersanne | i think it makes sense to make it the internal default, and if _DATA is null then we set it to false
14:42:54 Gundersanne | and see where we end up
14:43:01       pitti | oh, why false?
14:43:10 Gundersanne | becase otherwise the check-multi-machines ones fail
14:43:13       pitti | -  if (g_strcmp0 (options->knownhosts_data, authorize_knownhosts_data) == 0)
14:43:13       pitti | +  if (options->knownhosts_data == NULL || g_strcmp0 (options->knownhosts_data, authorize_knownhosts_data) == 0)
14:43:13       pitti |      options->knownhosts_authorize = TRUE;
14:43:18       pitti | I thought about doing this
14:43:35 Gundersanne | yea that would work
14:43:37 Gundersanne | i thikn
14:43:58           * | pitti runs local unit tests and throws it bot-wards
14:44:45       pitti | meh, of course unit tests fail
14:45:08       pitti | presumably because *these* are written with the thought in mind that the internal default is to not authorize
14:45:22       pitti | as these don't use the bridge manifest
14:45:41       pitti | this is quite silly - we run all unit tests against a cockpit-ssh config which never occurs in practice
14:46:16       pitti | ok, I have something to do now :)
14:46:59 Gundersanne | right
14:47:09 Gundersanne | yea the unit tests are not super representative of the real world uses
14:48:28       pitti | Gundersanne: but I think you already fixed these, so I'll steal from your changes :)
14:49:37       pitti | now this makes so much more sense already
14:49:56 Gundersanne | i did?
14:50:00       pitti | Gundersanne: and thanks for bearing with me :)
14:50:02 Gundersanne | on in 10143
14:50:16 Gundersanne | nah no worries pitti i think rubberducking each other makes us understand
14:50:21           * | Gundersanne quacks for good measure
14:50:50 Gundersanne | so gladly
14:51:21       pitti | not sure if the fix in 10143 is directly applicable, but I'll see
14:51:43 Gundersanne | so pitti _DATA is null when started from cockpit/ws
14:51:50 Gundersanne | it ignores the manifest entirely
14:51:54       pitti | right
14:52:07       pitti | and I figure that might be the "magic" switch between the two
14:52:09 Gundersanne | yea
14:52:35 Gundersanne | because cockpit-ssh currently needs to konw if it's started from cockpit/ws or via cockpitpeer
14:53:46       pitti | so you figure it might make more sense to just rename it to CCOCKPIT_SSH_KNOWN_HOSTS_AUTHORIZE=1 in the manifest?
14:53:58       pitti | manifest+bridge+c-ssh update in lockstep
14:54:02       pitti | c-ws doesn't
14:54:38       pitti | it's still an absolutely awful design, IMHO
14:55:01       pitti | (steering this through the protocol and relying on env vars like that)
```

File: src/ssh/cockpitsshoptions.c [hunk 1]
```diff
@@ -78,40 +75,43 @@ set_environment_bool (gchar **env,
...
   options->remote_peer = get_environment_val (env, "COCKPIT_REMOTE_PEER", "localhost");
+  options->connect_to_unknown_hosts = get_connect_to_unknown_hosts (env);
 
-  if (options->knownhosts_data != NULL && !options->knownhosts_authorize)
-    options->allow_unknown_hosts = TRUE;
-  else
-    options->allow_unknown_hosts = get_allow_unknown_hosts (env);
+  options->challenge_unknown_host_preconnect =
+      get_environment_bool (env, "COCKPIT_SSH_CHALLENGE_UNKNOWN_HOST_PRECONNECT", FALSE);
+  /* Fallback to _DATA */
```
Comment (martinpitt) [inline src/ssh/cockpitsshoptions.c]: This should get a "HACK:" comment that _DATA is obsolete, and will be removed at some point.

File: src/ssh/cockpitsshrelay.c [hunk 2]
```diff
@@ -531,6 +508,29 @@ static void cleanup_knownhosts_file (void)
...
+      if (value && value[0] == '\0')
+        ret = g_strdup ("* invalid key");
+      else
+        ret = g_strdup (value);
+    }
+
+
+  g_free (response);
+  return ret;
+}
```
Comment (martinpitt) [inline src/ssh/cockpitsshrelay.c]: Why was this moved further down? This might just be accidental diff noise?

Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Oh yea, probably noise; sorry will get that.

File: src/ssh/cockpitsshoptions.c (hunk 1 above)
Comment (Gundersanne) [reply src/ssh/cockpitsshoptions.c]: Will do

**Comment by martinpitt** [comment]
It now looks like we are approaching zero regressions \o/ (and it only took four (wo)manweeks to understand and untangle all this stuff!)

@stefwalter , would you mind having a look at this, at least at the rationale in the commit message and the documentation updates? also at [the summary above](https://github.com/cockpit-project/cockpit/pull/10143#issuecomment-425949732)?

**Comment by martinpitt** [comment]
This works now. We also have backwards compat shims for _DATA and _ALLOW_UNKNOWN, so this looks good to me. I'd still appreciate if @stefwalter could quickly sanity check this -- we talked about it last week, but a comit changelog is more comprehensive.

**Comment by Gundersanne** [comment]
I think i accidentally removed the "/ssh-bridge/knownhost-invalid" test, because it was between two authorize tests. So i think we can restore this one, it also succeeds.

**Comment by martinpitt** [comment]
@Gundersanne: ack, let's do that then. Thanks for spotting!

File: doc/authentication.md [hunk 3]
```diff
@@ -133,7 +133,7 @@ The following environment variables are set by cockpit-ws when spawning an auth
...
  * **COCKPIT_SSH_KNOWN_HOSTS_FILE** Path to knownhost files. Defaults to ```PACKAGE_SYSCONF_DIR/ssh/ssh_known_hosts```
- * **COCKPIT_SSH_KNOWN_HOSTS_DATA** Known host data to validate against or '*' to skip validation```
  * **COCKPIT_SSH_BRIDGE_COMMAND** Command to launch after a ssh connection is established. Defaults to ```cockpit-bridge``` if not provided.
+ * **COCKPIT_SSH_CHALLENGE_UNKNOWN_HOST_PRECONNECT** Set to ```1``` to receive an authorize challenge if the host is unknown, before connecting to the host. This is done in cockpit-bridge, but not for direct URL logins in cockpit-ws. This is an internal hack to mimic the obsolete magic ```COCKPIT_SSH_KNOWN_HOSTS_DATA=authorize``` mode, until cockpit-ws starts to set this through the JSON protocol. Don't set this outside of Cockpit, this is *not* API!
```
Comment (stefwalter) [inline doc/authentication.md]: If this is not API then why is are we documenting it? Please move this to a comment in the code instead.

File: doc/authentication.md [hunk 4]
```diff
@@ -133,7 +133,7 @@ The following environment variables are set by cockpit-ws when spawning an auth
 
 The following environment variables are used to set options for the ```cockpit-ssh``` process.
 
- * **COCKPIT_SSH_ALLOW_UNKNOWN**` Set to ```1``` to  allow connecting to hosts that are not saved in the current knownhosts file. If not set cockpit will only connect to unknown hosts if either the remote_peer is local or if the ```Ssh-Login``` section in ```cockpit.conf``` has a ```allowUnknown``` option set to a truthy value (```1```, ```yes``` or ```true```).
+ * **COCKPIT_SSH_CONNECT_TO_UNKNOWN_HOSTS**` Set to ```1``` to  allow connecting to hosts that are not present in the current ```known_hosts``` files. If not set, Cockpit will only connect to unknown hosts if either the remote peer is local, or if the ```Ssh-Login``` section in ```cockpit.conf``` has an ```allowUnknown``` option set to a true value (```1```, ```yes``` or ```true```).
```
Comment (stefwalter) [inline doc/authentication.md]: The 'allowUnknown' here should change to 'connectToUnknownHosts'

File: src/ssh/cockpitsshoptions.c [hunk 5]
```diff
@@ -78,40 +75,43 @@ set_environment_bool (gchar **env,
 }
 
 static gboolean
-get_allow_unknown_hosts (gchar **env)
+get_connect_to_unknown_hosts (gchar **env)
 {
   const gchar *remote_peer = g_environ_getenv (env, "COCKPIT_REMOTE_PEER");
 
+  /* Fallback to allowUnknown  */
+  if (!cockpit_conf_string (COCKPIT_CONF_SSH_SECTION, "connectToUnknownHosts") &&
```
Comment (stefwalter) [inline src/ssh/cockpitsshoptions.c]: Why is this cockpit_conf_string? Needs a comment at least. Because we use cockpit_conf_bool on the option below.

File: src/ssh/cockpitsshrelay.c [hunk 6]
```diff
@@ -705,15 +684,31 @@ set_knownhosts_file (CockpitSshData *data,
...
-      problem = "unknown-host";
-      goto out;
+      authorize_knownhosts_data = challenge_for_knownhosts_data (data);
+      if (authorize_knownhosts_data)
+        {
+          if (write_tmp_knownhosts_file (data, authorize_knownhosts_data, &problem))
+            {
+              host_known = cockpit_is_host_known (tmp_knownhost_file, host, port);
+              if (host_known)
+                data->ssh_options->knownhosts_file = tmp_knownhost_file;
```
Comment (stefwalter) [inline src/ssh/cockpitsshrelay.c]: What would cause the case where ```!host_known```? What would happen in that case? At the very least those answers should be commented here.

File: src/ssh/test-sshbridge.c [hunk 7]
```diff
@@ -182,10 +182,14 @@ setup_env (const TestFixture *fix)
...
   if (fix && fix->allow_unknown)
     {
-      env = g_environ_setenv (env, "COCKPIT_SSH_ALLOW_UNKNOWN",
+      env = g_environ_setenv (env, "COCKPIT_SSH_CONNECT_TO_UNKNOWN_HOSTS",
                               "true", TRUE);
     }
 
+  if (fix && fix->challenge_unknown_host_preconnect)
+    env = g_environ_setenv (env, "COCKPIT_SSH_CHALLENGE_UNKNOWN_HOST_PRECONNECT",
+                            "true", TRUE);
```
Comment (stefwalter) [inline src/ssh/test-sshbridge.c]: Need braces around a multiline (even if just wrapping).

File: src/ws/mock-config/cockpit/cockpit-alt.conf [hunk 8]
```diff
@@ -4,7 +4,7 @@ ProtocolHeader = X-Forwarded-Proto
 [Ssh-Login]
 command = mock-auth-command
 host = default-host
-allowUnknown = true
+connectToUnknownHosts = true
```
Comment (stefwalter) [inline src/ws/mock-config/cockpit/cockpit-alt.conf]: We should continue to test the old config name and environment variable name if we have code to support it. So lets copy this and create additional tests.

File: src/ssh/cockpitsshoptions.c (hunk 5 above)
Comment (martinpitt) [reply src/ssh/cockpitsshoptions.c]: This is to check whether the option is set at all. _bool cannot tell the difference between "unset" and "set to false".

Comment (stefwalter) [reply src/ssh/cockpitsshoptions.c]: Makes sense. Lets add to the comment there.

File: src/ssh/cockpitsshrelay.c (hunk 6 above)
Comment (martinpitt) [reply src/ssh/cockpitsshrelay.c]: This is fairly similar to the case above when we query sssd. An unknown host indeed Should Not Happen™ here, I'll add a warning and a comment similar to the sssd case.

**Comment by martinpitt** [comment]
Thanks @stefwalter for the detailled review! I pushed a fixup commit with all the corresponding changes. I also force-pushed the main commit, but only to update its commit log, and rebase to current master. The "check external projects" TODO items are still relevant.

**Comment by martinpitt** [comment]
I checked ovirt-{engine,node} with `git grep -i cockpit.ssh` and `git grep allowUnknown`. ovirt doesn't seem to use cockpit-ssh directly at all.

**Comment by martinpitt** [comment]
ManageIQ has a [cockpit-auth-miq](https://github.com/ManageIQ/manageiq/blob/master/tools/cockpit/cockpit-auth-miq) tool which uses `$COCKPIT_SSH_ALLOW_UNKNOWN`. We have backwards compat code for that, including tests, so this seems fine. This tool also uses `$COCKPIT_SSH_SUPPORTS_HOST_KEY_PROMPT`, but this was already removed 1½ years ago in commit 0a18c73650f6aa92478fa.

**Comment by stefwalter** [comment]
Lets talk with @mareklibra to make sure we're not breaking oVirt.

File: src/ws/mock-config/cockpit/cockpit-deprecated.conf [hunk 9]
```diff
@@ -0,0 +1,10 @@
+[WebService]
+ProtocolHeader = X-Forwarded-Proto
+
+[Ssh-Login]
+command = mock-auth-command
+host = default-host
+allowUnknown = true
```
Comment (stefwalter) [inline src/ws/mock-config/cockpit/cockpit-deprecated.conf]: Could you add a comment here that indicates this is the deprecated option being tested?

Comment (Gundersanne) [reply src/ws/mock-config/cockpit/cockpit-deprecated.conf]: Done!

**Comment by Gundersanne** [comment]
oki, spinning tests again and then we're good to go.

**Comment by martinpitt** [comment]
Remaining two failures are unrelated. Landing at last \o/

**Comment by Gundersanne** [comment]
@martinpitt  I must've retriggered f29 at least 15 times over the weekend + your retry loop to try to get it green. I think I'll spend some time on fixing check-multi-machines tomorrow :)
